## Q6. Euclidean and Manhattan Distance

---

### 🔍 এই Question-এ কী চাওয়া হয়েছে?

User 0 ও User 1-এর জন্য **2D vector** তৈরি করতে হবে এই দুটো feature দিয়ে:
- `income_std` (Q4.2-তে তৈরি করা standardized column)
- `daily_app_opens`

তারপর এই দুটো vector-এর মধ্যে:

| Task | কী করতে হবে |
|---|---|
| **Q6.1** | Vector তৈরি করা (already done) |
| **Q6.2** | **Euclidean distance** হিসাব করা |
| **Q6.3** | **Manhattan distance** হিসাব করা |
| **Q6.4** | কোনটি বড় এবং কেন — short answer |

### 🧠 দুটো Formula:

**Euclidean:**
$$d_E = \sqrt{(v1_1 - v2_1)^2 + (v1_2 - v2_2)^2}$$

**Manhattan:**
$$d_M = |v1_1 - v2_1| + |v1_2 - v2_2|$$

### Actual Vectors (এই dataset-এ):
- **v1** (User 0): `[0.9447, 48.0]`
- **v2** (User 1): `[-0.3246, 7.0]`


### Q6.1 — 2D Vectors তৈরি করা (Already Provided)

In [16]:
import pandas as pd
import numpy as np

# Now we'll load the data and rebuild income_std from Q4
df = pd.read_csv("02. Provided - module123_data.csv")

# Now we'll standardize the income column. It's important to do the standardization before building the vector.
col_mean = df["monthly_income"].mean()
col_std  = df["monthly_income"].std()
df["income_std"] = (df["monthly_income"] - col_mean) / col_std
# Here we've rebuilt the standardized income column from Q4.

# Now we'll build 2D vectors for user 0 and user 1
vec_cols = ["income_std", "daily_app_opens"]

v1 = df.loc[0, vec_cols].values.astype(float)
v2 = df.loc[1, vec_cols].values.astype(float)
# Here we've extracted float vectors for the first two users.

print("v1:", v1.round(4))
print("v2:", v2.round(4))


v1: [ 0.9447 48.    ]
v2: [-0.3246  7.    ]


`df.loc[0, vec_cols].values` → row 0 থেকে দুটো feature-এর value নিয়ে NumPy array তৈরি।
`.astype(float)` → **এটি জরুরি** — mixed column থেকে array নেওয়ার সময় dtype `object` হতে পারে, `float`-এ convert না করলে `.round()` বা arithmetic কাজ করে না।

**v1** = User 0: income_std=0.9447, daily_app_opens=48.0
**v2** = User 1: income_std=−0.3246, daily_app_opens=7.0


### Q6.2 — Euclidean Distance [5 Marks]

In [2]:
# Now we'll compute the Euclidean distance step by step, between v1 and v2
diff     = v1 - v2
sq_diff  = diff ** 2
euc_dist = np.sqrt(sq_diff.sum())
# Here we've calculated: subtract → square → sum → square root.

print("Step by step:")
print(f"  v1 - v2            = {diff.round(4)}")
print(f"  (v1 - v2)^2        = {sq_diff.round(4)}")
print(f"  Sum of squares     = {sq_diff.sum():.4f}")
print(f"  Euclidean distance = {euc_dist:.4f}")


Step by step:
  v1 - v2            = [ 1.2693 41.    ]
  (v1 - v2)^2        = [1.6112e+00 1.6810e+03]
  Sum of squares     = 1682.6112
  Euclidean distance = 41.0196


`v1 - v2` → element-wise বিয়োগ।
`** 2` → প্রতিটি পার্থক্যকে বর্গ করা।
`np.sqrt()` → বর্গমূল নেওয়া।
**Euclidean distance = 41.0196**


### Q6.3 — Manhattan Distance [5 Marks]

In [17]:
# Now we'll compute the Manhattan distance step by step
abs_diff = np.abs(v1 - v2)
man_dist = abs_diff.sum()
# Here we've calculated: subtract → absolute value → sum.

print("Step by step:")
print(f"  v1 - v2            = {(v1 - v2).round(4)}")
print(f"  |v1 - v2|          = {abs_diff.round(4)}")
print(f"  Manhattan distance = {man_dist:.4f}")


Step by step:
  v1 - v2            = [ 1.2693 41.    ]
  |v1 - v2|          = [ 1.2693 41.    ]
  Manhattan distance = 42.2693


`np.abs()` → negative হলেও positive করে দেয়।
`.sum()` → absolute difference গুলো যোগ করে।
**Manhattan distance = 42.2693**


### দুটো Distance পাশাপাশি দেখা

In [4]:
# Now we'll print a clean comparison of both distances
print("=" * 35)
print("  Distance Comparison")
print("=" * 35)
print(f"  Euclidean : {euc_dist:.4f}")
print(f"  Manhattan : {man_dist:.4f}")
print(f"  Larger    : {'Manhattan' if man_dist > euc_dist else 'Euclidean'}")
print("=" * 35)
# Here we've printed both results side by side for easy comparison.


  Distance Comparison
  Euclidean : 41.0196
  Manhattan : 42.2693
  Larger    : Manhattan


**Result:** Manhattan (42.27) > Euclidean (41.02)


### Q6.4 — Short Answer [5 Marks]

> **Question:** Which one is larger in your result, Euclidean or Manhattan distance and why does that usually happen based on their formulas?

---

**Answer:**

> In this result, the Manhattan distance (42.27) is larger than the Euclidean distance (41.02). This is typical because Euclidean distance takes a straight-line shortcut between two points by squaring the differences and then taking the square root, which compresses the total value. Manhattan distance, on the other hand, adds up the absolute differences along each dimension separately — like walking along a grid — so it always takes a longer or equal path compared to the straight line, making it larger or equal to the Euclidean distance in most cases.

> এই ফলাফলে Manhattan distance (৪২.২৭) Euclidean distance (৪১.০২)-এর চেয়ে বড়। এটাই স্বাভাবিক, কারণ Euclidean distance দুটি বিন্দুর মধ্যে সরাসরি সরলরেখায় যায় — পার্থক্যগুলো বর্গ করে তারপর বর্গমূল নেয়, যা মোট মানটাকে সংকুচিত করে ফেলে। অন্যদিকে Manhattan distance প্রতিটি dimension-এ আলাদাভাবে absolute পার্থক্যগুলো যোগ করে — অনেকটা grid-এর উপর দিয়ে হাঁটার মতো — তাই এটি সবসময় সরলরেখার চেয়ে লম্বা বা সমান পথ নেয়, ফলে বেশিরভাগ ক্ষেত্রে এটি Euclidean distance-এর চেয়ে বড় বা সমান হয়।

---